src/os/geo/jsts.js

```diff
@@ -526,6 +526,198 @@ os.geo.jsts.toPolygon = function(geometry) {
+ * @private
+ */
+os.geo.jsts.polygonize_ = function(geometry) {
+  var lines = jsts.geom.util.LineStringExtracter.getLines(geometry);
+  var polygonizer = new jsts.operation.polygonize.Polygonizer();
+  polygonizer.add(/** @type {jsts.Collection<jsts.geom.Geometry>} */ (lines));
+  var polys = polygonizer.getPolygons();
+  return jsts.geom.GeometryFactory.toPolygonArray(polys);
+};
+
+
+/**
+ * @param {ol.geom.Polygon|ol.geom.MultiPolygon} polygon
+ * @param {ol.geom.LineString} line
+ * @return {ol.geom.Polygon|ol.geom.MultiPolygon}
+ */
+os.geo.jsts.splitPolygonByLine = function(polygon, line) {
+  if (polygon && line) {
+    var olp = os.geo.jsts.OLParser.getInstance();
+    var jstsPolygon = olp.read(polygon);
+    var jstsLine = olp.read(line);
+
+    if (jstsPolygon && jstsLine) {
+      var nodedLinework = jstsPolygon.getBoundary().union(jstsLine);
+      var polys = os.geo.jsts.polygonize_(nodedLinework);
+
+      // only keep polygons which are inside the input
+      polys = polys.filter(function(poly) {
+        return jstsPolygon.contains(poly.getInteriorPoint());
+      });
+
+      polys = /** @type {Array<ol.geom.Polygon>} */ (polys.map(olp.write.bind(olp)));
+
+      var rings = [];
+      for (var i = 0, n = polys.length; i < n; i++) {
+        rings = rings.concat(polys[i].getCoordinates());
+      }
+
+      if (polys.length > 1) {
+        var multi = new ol.geom.MultiPolygon([]);
+        multi.setPolygons(polys);
+        return multi;
+      } else if (polys.length > 0) {
+        return polys[0];
+      }
+    }
+  }
+
+  return polygon;
+};
+
+
+/**
+ * @param {Array<ol.geom.Polygon|ol.geom.MultiPolygon>} polygons
+ * @private
+ */
+os.geo.jsts.flattenPolys_ = function(polygons) {
+  // do not cache length here as it is changing
+  for (var i = 0; i < polygons.length; i++) {
+    var item = polygons[i];
+    if (item.getType() === ol.geom.GeometryType.MULTI_POLYGON) {
+      var args = [i, 1].concat(/** @type {ol.geom.MultiPolygon} */ (item).getPolygons());
+      Array.prototype.splice.apply(polygons, args);
+    }
+  }
+};
+
+
+/**
+ * @param {ol.geom.Polygon} polygon
+ * @return {boolean}
+ */
+os.geo.jsts.needsSplit = function(polygon) {
+  var coords = polygon.getCoordinates()[0];
+  var proj = ol.proj.get(/** @type {string} */ (polygon.get(os.geom.GeometryField.PROJECTION)) ||
+    os.map.PROJECTION);
+
+  if (os.geo2.isPolarRing(coords, proj)) {
+    return true;
+  }
+
+  var extent = proj.getExtent();
+  var halfWidth = (extent[2] - extent[0]) / 2;
+
+  for (var i = 1, n = coords.length; i < n; i++) {
+    if (Math.abs(coords[i][0] - coords[i - 1][0]) > halfWidth) {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+
+proj4.defs('EPSG:3031',
+    '+proj=stere +lat_0=-90 +lat_ts=-71 +lon_0=0 +k=1 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs');
+proj4.defs('EPSG:3413',
+    '+proj=stere +lat_0=90 +lat_ts=70 +lon_0=-45 +k=1 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs');
```

Comment:
Placement of these seems random. Do we want to keep these here, or are we loading them from settings?

Reply:
They must exist in `proj4` regardless of whether they are configured in settings for use by the user.